keps/sig-network/20200204-cni-verification-rearchitecture.md

```diff
@@ -0,0 +1,622 @@
+
+### Goals
+
+- Rearchitect the way we write and define CNI NetworkPolicy test verifications
+- Increase the visibility and quality of documentation available for network policies
+
+
+### Non-goals
+
+- Make tests specific to CNI providers
+- Change NetworkPolicy APIs
+ 
+### Related issues
+
+As an overall improvement, this KEP will help to address the solutions for several existing issues in upstream Kubernetes.  Some of these issues have been duct-taped upstream, but our overarching goal is to reduce the amount of work required to verify that any such issues have been properly addressed and accounted for in the documentation, testing, and semantic aspects of how the API for NetworkPolicy itself is defined.
+
+- https://github.com/kubernetes/kubernetes/issues/87857 (docs and understandability)
+- https://github.com/kubernetes/kubernetes/issues/87893 (holes in our test coverage matrix)
+- https://github.com/kubernetes/kubernetes/issues/85908 (failing tests, unclear semantics)
+- https://github.com/kubernetes/kubernetes/issues/86578 (needs e2e coverage)
+- https://github.com/kubernetes/kubernetes/issues/87709 (logging of netpol actions, will help describing states we reach) 
+- https://github.com/projectcalico/felix/issues/2032 (non-deterministic time frames for policy applications - addressable through published performance tests which measure time to policy implementation for several pods in several namespaces.)
+- https://github.com/projectcalico/felix/issues/2008 (need to test postStart pods in networkpolicy upstream.)
+- https://github.com/vmware-tanzu/antrea/issues/381 (addressable by performance tests *or* node-targeted tests.)
+- https://github.com/kubernetes/kubernetes/issues/88375 (the test matrix for Egress is almost entirely empty, decreasing the verbosity of new tests will organically increase likelihood of new test submissions over time.)
+
+
+### Consequences of this problem
+ 
+The consequences of this problem is that
+ 
+- CNI providers cannot easily be compared for functionality.
+- CNI providers implementing network policies must carry a lot of downstream test functionality.
+- Testing a CNI provider for Kubernetes compatibility requires a lot of interpretation and time investment.
+- Extending NetworkPolicy tests is time consuming and error prone, without a structured review process and acceptance standard.
+- It is hard to debug tests, due to the performance characteristics - pods are deleted after each test, so we cannot reproduce the state of the cluster easily.
+
+## A high level outline of Pod Traffic Pathways
+
+Before diving into test details, we outline different types of pod traffic.  Each one of these types of traffic may pose different bugs to a CNI provider.  
+
+Intranode
```

Comment:
I was actually thinking of deleting this section , unless we have specific expectations from it, @bowei  --- i feel like it might not be essential to the overall purpose of the KEP, as you mentioned in your original comment....  Would that be ok ?